File: summary/ruby/17_shakutori.rb
def shakutori(arr, k)
  n = arr.length
  res = 0
  right = 0
  sum = 0

  (0...n).each do |left|
    while right < n && sum + arr[right] <= k
      sum += arr[right]
      right += 1
    end

    res += right - left

    if right == left
      right += 1
    else
      sum -= arr[left]
    end
  end

  res
end

# Example usage:
input_array = gets.chomp.split.map(&:to_i)
k = gets.to_i
result = shakutori(input_array, k)
puts "Number of subarrays with sum <= #{k}: #{result}"

File: summary/go/17_shakutori.go
package main

import "fmt"

func shakutori(arr []int, k int) int {
	n := len(arr)
	res := 0
	right := 0
	sum := 0

	for left := 0; left < n; left++ {
		for right < n && sum+arr[right] <= k {
			sum += arr[right]
			right++
		}

		res += right - left

		if right == left {
			right++
		} else {
			sum -= arr[left]
		}
	}

	return res
}

// Example usage:
func main() {
	inputArray := []int{1, 2, 3, 4, 5}
	k := 5
	result := shakutori(inputArray, k)
	fmt.Printf("Number of subarrays with sum <= %d: %d\n", k, result)
}
